src/bin/rational.rs

```diff
@@ -175,27 +175,139 @@ impl SynthLanguage for Math {
     /// Term enumeration.
     fn make_layer(synth: &Synthesizer<Self>, _iter: usize) -> Vec<Self> {
         let mut to_add = vec![];
+
+        log::info!("Previous layer: {} enodes, {} eclasses ",
+                    synth.egraph.total_number_of_nodes(),
+                    synth.egraph.number_of_classes());
+
         for i in synth.ids() {
             for j in synth.ids() {
-                if synth.egraph[i].data.exact && synth.egraph[j].data.exact {
-                    continue;
+                // must have at least one non-constant term
+                // and at least one term from previous iteration
+                if synth.egraph[i].data.gen + 1 == _iter || synth.egraph[j].data.gen + 1 == _iter {
+                    if _iter > synth.params.no_constants_above_iter {
+                        if synth.egraph[i].data.exact || synth.egraph[j].data.exact {
+                            continue;
+                        }
+                    } else {
+                        if synth.egraph[i].data.exact && synth.egraph[j].data.exact {
+                            continue;
+                        }
+                    };
+
+                    to_add.push(Math::Add([i, j]));
+                    // to_add.push(Math::Sub([i, j]));
+                    to_add.push(Math::Mul([i, j]));
+                    // to_add.push(Math::Div([i, j]));
                 }
-                to_add.push(Math::Add([i, j]));
-                to_add.push(Math::Sub([i, j]));
-                to_add.push(Math::Mul([i, j]));
-                to_add.push(Math::Div([i, j]));
             }
-            if synth.egraph[i].data.exact {
-                continue;
+
+            // must have at least one non-constant term
+            // and at least one term from previous iteration
+            if !synth.egraph[i].data.exact && synth.egraph[i].data.gen + 1 == _iter  {
+                to_add.push(Math::Abs(i));
+                to_add.push(Math::Neg(i));
             }
-            to_add.push(Math::Abs(i));
-            to_add.push(Math::Neg(i));
         }
 
         log::info!("Made a layer of {} enodes", to_add.len());
         to_add
     }
 
+    fn valid_constants(
```

Comment:
I haven't looked at the entire PR yet, but this seems tricky to me ---  traversing the egraph recursively can be very expensive so we should discuss if we really need this or there is a way to accomplish similar effects differently.

I will comment more on the PR later. Looks fantastic so far!!